compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -769,16 +769,45 @@ object PatternMatcher {
     }
 
     /** Emit cases of a switch */
-    private def emitSwitchCases(cases: List[(List[Tree], Plan)]): List[CaseDef] = (cases: @unchecked) match {
-      case (alts, ons) :: cases1 =>
-        val pat = alts match {
-          case alt :: Nil => alt
-          case Nil => Underscore(defn.IntType) // default case
-          case _ => Alternative(alts)
+    private def emitSwitchCases(cases: List[(List[Tree], Plan)]): List[CaseDef] = cases.foldLeft((List[CaseDef](), List[Tree]())) {
```

Comment:
Generally our code style is to prefer tail-recursive formulation over folds and to replace `Option` with sentinels. This generally increases clarity and avoids unnecessary allocations. Allocations are your enemy - they make the compiler slow, more than anything else. And since they are spread out, there are no hot spots to address, you really have to take care to avoid them everywhere as much as is reasonable.

So would prefer the old formulation of `emitCases` over the new one.